Push/Autopush.swift

```diff
@@ -0,0 +1,141 @@
+    ///  - Returns: A `SubscriptionResponse` that includes:
+    ///         - Encryption keys to be used to encrypt any push notifications that will be sent to this device
+    ///         - A URL that consumer can use to send push notifications to this device
+    ///  - Throws: If the underlying native call to subscribe with the autopush servers fails
+    func subscribe(scope: String) async throws -> SubscriptionResponse
+
+    /// Unsubscribes a push subscription with the given scope.
+    ///
+    /// - Parameter scope: A consumer controlled string. When push notifications are decrypted, the scope will be broadcased so consumers
+    ///                                             can handle the notification
+    /// - Returns: `true` if the subscription was unsubscribed, `false` if the subscription did not exist
+    /// - Throws: If the underlying native call to unsubscribe with the autopush servers fails
+    func unsubscribe(scope: String) async throws -> Bool
+
+    /// Unsubscribes from all scopes
+    ///
+    /// - Throws: If the underlying native call to unsubscribe with the autopush servers fails
+    func unsubscribeAll() async throws
+
+    /// Decrypts an incoming push payload from `Autopush` server
+    ///
+    /// - Parameter payload: A map of String keys and String values representing payload as sent by the push servers
+    ///
+    /// - Returns: `DecryptResponse`, which includes both the decrypted payload, and the scope the push notification was for
+    /// - Throws: If the native push client was unable to decrypt the payload
+    func decrypt(payload: [String: String]) async throws -> DecryptResponse
+}
+
+open class Autopush {
+    private var pushClient: PushManagerProtocol?
+    private let dbPath: String
+
+    public init(dbPath: String) {
+        self.dbPath = dbPath
+        self.pushClient = nil
+    }
+
+    /// Suspends execution and runs the given function in the global default dispatch queue against a native push client
+    ///
+    /// - Parameters:
+    ///   - fn:  A function that takes a `PushManagerProtocol` and returns the value `withClient` would return
+    ///
+    /// - Returns: The return value of `fn`
+    ///
+    /// - Throws:
+    ///     - if `withClient` could not get an opened push connection
+    ///     - An error was throwing from `fn`
+    ///
+    /// - Note:
+    ///     - `withClient` assumes that `fn` might need to do blocking IO and proactively schedules it on the global default dispatch queue
+    ///     - `withClient` will attempt to open the underlying native client if not open already - The openning is done using `getOpenedPushClient`
+    private func withClient<T>(_ fn: @escaping (PushManagerProtocol) throws -> T) async throws -> T {
+        let pushClient = try await getOpenedPushClient()
+        return try await withCheckedThrowingContinuation { continuation in
+            DispatchQueue.global().async {
```

Comment:
I'm not quite sure what I'm doing here is advised or not. TLDR:
- `fn` might included some I/O so we want to make sure it's not on the main thread
- I wasn't sure if I should "just" make this API async and leave the threading of it to the consumer as that seems risky that a consumer would trigger it on the main thread, but if the swift concurrency way of doing this is do that then I'm happy to change it

Reply:
Because this is flagged as an async function and is not flagged as @MainActor it will not run on the main thread, so you do not need to add the continuation or the DispatchQueue call. This pattern really simplifies how you deal with threading. 
I'll try to have that presentation on the topic ready soon, hoping to have it ready by end of week and I'll present it sometime next week.

Reply:
That does simplify things quite a bit 😮